In [2]:
using BenchmarkTools
n = 20_000
usnp = rand(Int16, n, n)
usnp[1:3, 1:5]

3×5 Matrix{Int16}:
  6186  12178    4836   -5893    842
 27408  32118  -11376   13984   5149
 -3103   6081    -633  -11722  -4430

In [3]:
@benchmark b = Int8.(isodd.(usnp))

BenchmarkTools.Trial: 36 samples with 1 evaluation.
 Range (min … max):  121.054 ms … 177.605 ms  ┊ GC (min … max):  0.00% … 29.48%
 Time  (median):     123.904 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):   139.390 ms ±  24.617 ms  ┊ GC (mean ± σ):  11.12% ± 13.50%

  ▄█▆ ▂                                                       ▆  
  ███▆█▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▁▁▁▁▁▄██ ▁
  121 ms           Histogram: frequency by time          178 ms <

 Memory estimate: 381.47 MiB, allocs estimate: 7.

In [4]:
@benchmark begin
    b = zeros(Int8, n, n)
    b .&= usnp
end

BenchmarkTools.Trial: 15 samples with 1 evaluation.
 Range (min … max):  330.934 ms … 395.126 ms  ┊ GC (min … max): 0.00% … 13.38%
 Time  (median):     335.385 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   349.906 ms ±  24.341 ms  ┊ GC (mean ± σ):  3.68% ±  5.73%

  ██▁█▁  ▁ ▁              ▁               ▁           ▁     ▁ ▁  
  █████▁▁█▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁█▁█ ▁
  331 ms           Histogram: frequency by time          395 ms <

 Memory estimate: 381.47 MiB, allocs estimate: 4.

In [5]:
@benchmark b = Int8.(usnp .& 1)

BenchmarkTools.Trial: 14 samples with 1 evaluation.
 Range (min … max):  348.242 ms … 410.972 ms  ┊ GC (min … max): 0.00% … 12.95%
 Time  (median):     353.032 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   367.242 ms ±  25.485 ms  ┊ GC (mean ± σ):  3.82% ±  5.70%

   █  █                                                          
  ▇█▇▁█▇▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▇▁▇▇ ▁
  348 ms           Histogram: frequency by time          411 ms <

 Memory estimate: 381.47 MiB, allocs estimate: 7.

In [6]:
@benchmark b = Int8.(usnp .> 0)

BenchmarkTools.Trial: 29 samples with 1 evaluation.
 Range (min … max):  155.984 ms … 215.595 ms  ┊ GC (min … max): 0.00% … 24.97%
 Time  (median):     158.061 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   174.797 ms ±  25.616 ms  ┊ GC (mean ± σ):  9.26% ± 11.59%

  █▂                                                        ▂    
  ██▄▄▄▄▄▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▅▄ ▁
  156 ms           Histogram: frequency by time          216 ms <

 Memory estimate: 381.47 MiB, allocs estimate: 7.

In [7]:
@benchmark b = Int8.(usnp .& Int8(1))

BenchmarkTools.Trial: 15 samples with 1 evaluation.
 Range (min … max):  327.751 ms … 392.260 ms  ┊ GC (min … max): 0.00% … 13.96%
 Time  (median):     332.591 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   346.049 ms ±  23.833 ms  ┊ GC (mean ± σ):  3.82% ±  5.95%

  █  ▃                                                           
  █▇▇█▇▁▇▁▁▁▇▁▁▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇▁▇▁▁▁▁▇ ▁
  328 ms           Histogram: frequency by time          392 ms <

 Memory estimate: 381.47 MiB, allocs estimate: 7.

In [8]:
b = ones(Int8, n, n)
@benchmark b .&= usnp

BenchmarkTools.Trial: 21 samples with 1 evaluation.
 Range (min … max):  244.986 ms … 250.442 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     246.448 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   246.758 ms ±   1.553 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █         ▃   ▃ ▃             ▃                                
  █▁▇▇▁▇▁▁▁▁█▁▁▁█▁█▁▇▁▁▁▁▁▁▇▇▁▁▁█▁▇▁▁▁▁▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▇▁▁▁▁▁▇ ▁
  245 ms           Histogram: frequency by time          250 ms <

 Memory estimate: 64 bytes, allocs estimate: 2.

## Conclusion

| Method | Best time (ms) | Rank |
| --- | :---: | :---: |
| `isodd.(mat)` | 121 | 1 |
| `.> 0` | 156 | 2 |
| `.&=1` | 245 | 3 |
| `.& Int8(1)` | 328 | 4 |
| `ones(Int8) .&= mat` | 331 | 5 |
| `Int8.(mat .& 1)` | 348 | 6 |